src/Npgsql/Internal/PgReader.cs

```diff
@@ -544,43 +550,53 @@ internal void ThrowIfStreamActive()
     }
 
     internal bool CommitHasIO(bool resuming) => Initialized && !resuming && FieldRemaining > 0;
-    internal ValueTask Commit(bool async, bool resuming)
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal void Commit(bool resuming)
     {
         if (!Initialized)
-            return new();
+            return;
 
         if (resuming)
         {
             if (!Resumable)
                 ThrowHelper.ThrowInvalidOperationException("Cannot resume a non-resumable read.");
-            return new();
+            return;
         }
 
         // We don't rely on CurrentRemaining, just to make sure we consume fully in the event of a nested scope not being disposed.
         // Also shut down any streaming, pooled arrays etc.
-        if (_requiresCleanup || FieldRemaining > 0)
-            return Slow(async);
+        if (_requiresCleanup || (!_fieldCompleted && FieldRemaining > 0))
+        {
+            CommitSlow();
+            return;
+        }
 
-        _fieldSize = default;
         _fieldStartPos = -1;
-        _resumable = false;
-        _fieldFormat = default;
-        if (_currentSize is not -1)
+        Debug.Assert(!Initialized);
+
+        // These will always be re-initialized by Init()
+        // _fieldSize = default;
+        // _fieldFormat = default;
+        // _resumable = default;
+        // _fieldCompleted = default;
+
+        if (HasCurrent)
         {
             _currentStartPos = 0;
             _currentBufferRequirement = default;
             _currentSize = -1;
+            Debug.Assert(!HasCurrent);
         }
-        Debug.Assert(!Initialized);
-        return new();
 
-        async ValueTask Slow(bool async)
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        void CommitSlow()
         {
             // Shut down any streaming and pooling going on on the column.
             if (_requiresCleanup)
             {
                 if (_userActiveStream is { IsDisposed: false })
-                    await DisposeUserActiveStream(async).ConfigureAwait(false);
+                    DisposeUserActiveStream(async: false).GetAwaiter().GetResult();
 
                 if (_pooledArray is not null)
```

Comment:
It seems like we call `Commit` every time we consume a row, which isn't a good thing for the pooled array (the case I have in mind is reading sequentially multiple rows with big strings, which results in constant rent/return).

Reply:
If we have such big strings the rent/return cost will be dwarfed by the decoding/potential IO. arraypool rent return is only slightly slower than an alloc + its gc cost

Reply:
Yeah, that's true, OK.